### DISCUSSION TITLE
How can I upload data in ldif (or any other)  format?

### DISCUSSION THREAD

**Discussion by gredwhite** [diegogslomp/samba-ad-dc#18] [Q&A] [answered]
I have ldif file and I want to upload it to  server ? Is there way to do it ? I tried to do it via samba-tool but no luck please help

**Comment by diegogslomp** [comment]
Hello

You can create a bind mount to upload the files, are you using the docker compose file or `docker run` command?

To bind to a existing container, commit a new image and run it again with the bind command like described [here](https://stackoverflow.com/questions/28302178/how-can-i-add-a-volume-to-an-existing-docker-container)

To bind using docker compose check [here](https://docs.docker.com/storage/bind-mounts/)

[Here](https://wiki.samba.org/index.php/Back_up_and_Restoring_a_Samba_AD_DC#Offline.2Flocal_DC_backup) you can check how to backup and restore a domain controller. Any trouble please let me know.

**Comment by gredwhite** [comment]
> Hello
> 
> You can create a bind mount to upload the files, are you using the docker compose file or `docker run` command?
> 
> To bind to a existing container, commit a new image and run it again with the bind command like described [here](https://stackoverflow.com/questions/28302178/how-can-i-add-a-volume-to-an-existing-docker-container)
> 
> To bind using docker compose check [here](https://docs.docker.com/storage/bind-mounts/)
For local development I use docker compose file but for test I use testContainers


Where should I put ldif file on container side to Samba read it during/after startup  ?

**Reply by diegogslomp** [reply]
The `/usr/local/samba` folder is inside a persistent storage, maybe you want to put inside `/usr/local/samba/share/setup` where other ldif files are located.

**Reply by gredwhite** [reply]
I have no samba folder inside **/usr/local** is it expected ?

![image](https://github.com/diegogslomp/samba-ad-dc/assets/6998637/a4244f2d-8459-4d99-a6d7-f4f76cb91439)

**Reply by diegogslomp** [reply]
If you are using the docker image from dockerhub you should see it, even before the domain provision:

```bash
~ docker run -it --rm diegogslomp/samba-ad-dc ls /usr/local
bin  etc  games  include  lib  lib64  libexec  samba  sbin  share  src
~ docker run -it --rm diegogslomp/samba-ad-dc ls /usr/local/samba
bin  bind-dns  etc  include  lib  lib64  libexec  private  sbin  share  var
~
```

**Reply by gredwhite** [reply]
Well - I will sort it out. Is it enough just put the ldif file to that location ? Will it be uploaded during samba startup ?

**Reply by diegogslomp** [reply]
Simply adding the file to the folder will not load it, the most efficient way that I know is to backup and restore a domain as described [here](https://wiki.samba.org/index.php/Back_up_and_Restoring_a_Samba_AD_DC#Offline.2Flocal_DC_backup). Any trouble let me know.

**Reply by gredwhite** [reply]
Sorry for potentially studpid follow up question but quickly looked at your link and found out the command for restore

`sudo samba-tool domain backup restore --backup-file=<tar-file> --newservername=<DC-name> --targetdir=<new-samba-dir>`


But I have no **tar** file. I have **ldif** file

**Reply by diegogslomp** [reply]
You need create a backup tar file from you previous domain controller to restore it to a new one. According to [this](https://wiki.samba.org/index.php/Administer_Unix_Attributes_in_AD_using_samba-tool_and_ldb-tools), maybe you can add data with something like:

```bash
ldbmodify -H /usr/local/samba/private/sam.ldb /tmp/user.ldif -U Administrator
```

**Reply by gredwhite** [reply]
To be honest I need a format which I will be able to modify like ldif. Source of backup is not an any existing DC.

**Reply by gredwhite** [reply]
So first of all I would like to know if it is possible to use ldif file as source for initial data upload to LDAP db of samba ?

And if answer - yes I want to know how to do it.

**Reply by gredwhite** [reply]
More details are here https://stackoverflow.com/questions/77224006/how-to-upload-ldif-file-to-samba-dc-server

**Reply by diegogslomp** [reply]
Ok so first try to create a new fresh domain and test it, use the steps from this repo README. After you have a funcional domain controler, you should have the `/usr/local/samba/private/sam.ldb` file created. After that to use ldapmodify with your file, change `/tmp/user.ldif` in the command. Print here the results.

```bash
ldbmodify -H /usr/local/samba/private/sam.ldb /tmp/user.ldif -U Administrator
```

**Comment by diegogslomp** [comment]
> More details are here https://stackoverflow.com/questions/77224006/how-to-upload-ldif-file-to-samba-dc-server

here you are trying to use a instantlinux/samba-dc:latest image that don't use the usr/local/samba but /var/lib/samba, the /usr/local/samba is created when samba is compiled from source

**Reply by gredwhite** [reply]
Do I have any option to solve my issue ?

**Reply by diegogslomp** [reply]
You can try creating a new fresh domain and ldapmodify as I described in the previous reply

**Comment by diegogslomp** [comment]
If you want to use the instalinux image, try to locate the sam.ldb, maybe in `/var/lib/samba/private/sam.ldb`  and use ldbmodify with your ldif:
ldbmodify -H /var/lib/samba/private/sam.ldb /tmp/yourfile.ldif -U Administrator

**Reply by gredwhite** [reply]
Well, it made be closer to success because this file really exists:

![image](https://github.com/diegogslomp/samba-ad-dc/assets/6998637/4f44ad37-82fb-411c-bc68-489f8541dd8f)


But when I run the command you suggested:

```
ldbmodify -H /var/lib/samba/private/sam.ldb /tmp/data.ldif -U Administrator
```
I get an error:
```
Invalid option -U: unknown option
Usage: ldbmodify <options> <ldif...>
Modifies a ldb based upon ldif change records

Usage: [OPTION...]
  -H, --url=URL                   database URL
  -b, --basedn=DN                 base DN
  -e, --editor=PROGRAM            external editor
  -s, --scope=SCOPE               search scope
  -v, --verbose                   increase verbosity
      --trace                     enable tracing
  -i, --interactive               input from stdin
  -r, --recursive                 recursive delete
      --modules-path=PATH         modules path
      --num-searches=INT          number of test searches
      --num-records=INT           number of test records
  -a, --all                       (|(objectClass=*)(distinguishedName=*))
      --nosync                    non-synchronous transactions
  -S, --sorted                    sort attributes
  -o OPTION                       ldb_connect option
      --controls=STRING           controls
      --show-binary               display binary LDIF
      --paged                     use a paged search
      --show-deleted              show deleted objects
      --show-recycled             show recycled objects
      --show-deactivated-link     show deactivated links
      --reveal                    reveal ldb internals
      --relax                     pass relax control
      --cross-ncs                 search across NC boundaries
      --extended-dn               show extended DNs

Help options:
  -?, --help                      Show this help message
      --usage                     Display brief usage message
```

I tried to remove -U option but I get an another error message:
```
/ # ldbmodify -H /var/lib/samba/private/sam.ldb /tmp/data.ldif
WARNING: Module [samba_dsdb] not found - do you need to set LDB_MODULES_PATH?
Unable to load modules for /var/lib/samba/private/sam.ldb: (null)
Failed to connect to /var/lib/samba/private/sam.ldb - (null)
```

**Reply by diegogslomp** [reply, marked as answer]
Try adding the LDB_MODULES_PATH to the command, change the lib64 to lib or check the correct path
```bash
LDB_MODULES_PATH=/usr/lib64/samba/ldb ldbmodify -H /var/lib/samba/private/sam.ldb /tmp/data.ldif
```

**Reply by gredwhite** [reply]
eventually this works(path a bit differs from yours)
```
LDB_MODULES_PATH=/usr/lib/samba/ldb/ ldbmodify -H /var/lib/samba/private/sam.ldb /tmp/singleUser.ldif 
```

Thank you!
